### PR TITLE
Add language packs automatic delivery code.

### PR DESCRIPTION
**Pull request by budzanowski**
# Automatic translations delivery code.
WooCommerce is leveraging WordPress Core automatic updates system to submit update information from woocommerce.com for WooCommerce plugins and themes. This PR further extends this system with the ability to automatically update language packs for plugins that are coming from `woocommerce.com` . This code cooperates with `translate.wordpress.com` GlotPress server. 

In short, the code collects information about which plugins from `woocommerce.com` are installed and active on the site. Then it checks if those plugins are activated for automatic translations ( filter on the plugin site ). Then using that information plus information about locales used on the site a request if formed and passed to `translate.wordpress.com` updates endpoint. Using returned information the `update_plugins` transient is populated. `update_plugins` transient is a part of WordPress core updates system. All updates are then performed without WooCommerce plugging interaction. 

### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Closes # .

### How to test the changes in this Pull Request:
Unfortunately, the translate.wordpress.com work is ongoing so to fully test this a wpcom  sandbox is required.

1. Use this branch.
2. Install and activate WooCommerce Bookings from this branch: https://github.com/woocommerce/woocommerce-bookings/tree/add/subscribe-to-automatic-translations . It has a filter that enables translations updates for this plugin.
3. Use wpcom sanbox with following patch D45269-code. Point following urls to your sandbox IP ( in /etc/hosts ): 
-- public-api.wordpress.com
-- wordpress.com
-- translate.wordpress.com
4. Everything is happening quite automatically but usually, a visit to the `wp-admin/update-core.php` triggers the update flow.
5. Change language to for example Spanish.
6. Do point 4 again.
7. The `/wp-content/languages/plugins` folder should have WooCommerce Bookings locale files:
image
This ^ is from my setup where I have Italian, German, and Spanish files.

This is a complicated setup. I am more than happy to help test this out or provide additional guidance.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### Changelog entry

> Enhancement - Added automatic language pack updates for WooCommerce.com extensions.

### CODE REVIEW

File: includes/admin/helper/class-wc-helper-updater.php [hunk 1]
```diff
@@ -162,6 +165,93 @@ public static function get_update_data() {
...
+				headers_sent() || WP_DEBUG ? E_USER_WARNING : E_USER_NOTICE
+			);
+			$raw_response = wp_remote_post( $http_url, $options );
+		}
+
+		if ( is_wp_error( $raw_response ) || 200 != wp_remote_retrieve_response_code( $raw_response ) ) {
+			return array();
+		}
+
+		$response = json_decode( wp_remote_retrieve_body( $raw_response ), true );
```
Comment (akirk) [inline includes/admin/helper/class-wc-helper-updater.php:246]: We could use this for hardcoding it for testing:
```suggestion
		$response = json_decode( '{"translations":[{"language":"de","version":"4.2.0","updated":"'+ date( 'Y-m-d H:i:s' ) . '","english_name":"German","native_name":"Deutsch","package":"https:\/\/translate.wordpress.com\/projects\/woocommerce\/woocommerce-bookings-availability\/stable\/de\/default\/export-translations\/?format=language-pack","iso":{"1":"de"}}]}', true );
```

File: includes/admin/helper/class-wc-helper-updater.php [hunk 2]
```diff
@@ -162,6 +165,93 @@ public static function get_update_data() {
...
+				'all'          => wp_json_encode( true ),
+			),
+			'user-agent' => 'WordPress/' . $wp_version . '; ' . home_url( '/' ),
+		);
+
+		if ( $extra_stats ) {
+			$options['body']['update_stats'] = wp_json_encode( $extra_stats );
+		}
+
+		$url      = 'htts://translate.wordpress.com/projects/';
```
Comment (beaulebens) [inline includes/admin/helper/class-wc-helper-updater.php:222]: `htts` is not a valid protocol ;)

Comment (budzanowski) [reply includes/admin/helper/class-wc-helper-updater.php:222]: It is a dummy placeholder. No worries. Good spotting!

File: includes/admin/helper/class-wc-helper-updater.php (hunk 1 above)
Comment (budzanowski) [reply includes/admin/helper/class-wc-helper-updater.php:236]: For testing, we can use a real endpoint https://code.a8c.com/D45269

Comment (budzanowski) [reply includes/admin/helper/class-wc-helper-updater.php:236]: We can use a real endpoint for testing D45269-code.

File: includes/admin/helper/class-wc-helper-updater.php [hunk 3]
```diff
@@ -162,6 +165,118 @@ public static function get_update_data() {
...
+		);
+
+		// Something wrong happened on the translate server side.
+		$response_code = wp_remote_retrieve_response_code( $raw_response );
+		if ( 200 !== $response_code ) {
+			return array();
+		}
+
+		// General error.
+		if ( is_wp_error( $raw_response ) ) {
```
Comment (rrennick) [inline includes/admin/helper/class-wc-helper-updater.php:241]: `wp_remote_retrieve_response_code()` returns `''` when the parameter is a `WP_Error` so this check shouldn't be necessary.

Comment (budzanowski) [reply includes/admin/helper/class-wc-helper-updater.php]: Cool. I have missed that.

**Comment by budzanowski** [comment]
Last commend updates the API URL, it was also updated in D45269-code so if anyone is using that they will need the latest changes.

**Review by jonathansadowski — COMMENTED** [review]
I attempted to test this by following the test plan. When I updated my language to Italian and visited `wp-admin/update-core.php`, I saw the following in my error log:

```
PHP Notice:  Undefined index: it_IT in /srv/www/wordpress-one/public_html/wp-content/plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php on line 258
```

and the return value from `WC_Helper_Updater::get_translations_update_data()` was an empty array. Once I manually updated the translations from `wp-admin/update-core.php`, that warning went away. Is that intended?

I also wasn't seeing any new language files in `wp-content/languages/plugins`, but I'm not sure if that was related to the former issue.

**Comment by budzanowski** [comment]
^ Fixed.

**Review by jonathansadowski — APPROVED** [review]
Thanks for the changes. The code makes sense to me, and I verified that after your recent change it works, without any warnings in the error log. I also verified that without D45269-code landed, it will gracefully do nothing. Nice work. 👍

**Review by ObliviousHarmony — APPROVED** [review]
Everything here works as expected!

**Pull request merged**